### PR TITLE
close #1991 add virtual waiting room API

### LINKED ISSUE #1991: Implement a virtual waiting room and a reserved timeslot
To implement a virtual waiting room and reserve time slots in Rails for handling high-load events, follow these steps:

### 1. **Set up the Virtual Waiting Room:**

A virtual waiting room helps manage traffic surges by queuing users during peak times.

#### Steps:
1. **Redis-based Queue:**
   - Use Redis as a fast in-memory store to queue users.
   - Implement a rate-limiting mechanism using `redis` gem to track the number of users allowed into the app simultaneously.
   
   ```ruby
   # Gemfile
   gem 'redis'
   
   # app/services/virtual_waiting_room_service.rb
   class VirtualWaitingRoomService
     def initialize(event_id)
       @event_id = event_id
       @redis = Redis.new
       @queue_key = "waiting_room:#{event_id}"
       @max_active_users = 100 # Set the max number of active users
     end
   
     def add_user_to_queue(user_id)
       @redis.lpush(@queue_key, user_id)
     end
   
     def allow_users_in
       active_users = @redis.lrange(@queue_key, 0, @max_active_users - 1)
       @redis.ltrim(@queue_key, @max_active_users, -1) # Keep remaining users in the queue
       active_users
     end
   end
   ```

2. **Session-based Limiting:**
   - Limit the number of users in the app at any time using sessions. If the limit is reached, users are placed in the waiting room.
   
   ```ruby
   # app/controllers/waiting_room_controller.rb
   class WaitingRoomController < ApplicationController
     def index
       if current_active_users_count < MAX_ACTIVE_USERS
         # Allow user into the event
         VirtualWaitingRoomService.new(event_id).allow_users_in
         redirect_to event_path(event_id)
       else
         # Show waiting room page
         render 'waiting_room/index'
       end
     end
   
     private
   
     def current_active_users_count
       # Check active users count in Redis or other tracking system
     end
   end
   ```

3. **Periodic User Admission:**
   - Periodically move users from the waiting room to the event using a background job or cron task. This can be done using `Sidekiq` or `Delayed Job` to manage the user flow.

   ```ruby
   class UserAdmissionJob
     include Sidekiq::Worker

     def perform(event_id)
       waiting_room_service = VirtualWaitingRoomService.new(event_id)
       allowed_users = waiting_room_service.allow_users_in

       allowed_users.each do |user_id|
         # Notify users via WebSocket or email that they are allowed in
         UserNotifier.send_admission_email(user_id)
       end
     end
   end
   ```

### 2. **Time Slot Reservation:**

Time slots can be used to reduce traffic at specific times by allowing users to reserve access during certain periods.

#### Steps:
1. **Create Time Slot Model:**
   - Each event can have multiple time slots that users can reserve.

   ```ruby
   # app/models/time_slot.rb
   class TimeSlot < ApplicationRecord
     belongs_to :event
     has_many :reservations

     validates :start_time, :end_time, presence: true
     validates :capacity, numericality: { greater_than_or_equal_to: 0 }

     def available_capacity
       capacity - reservations.count
     end
   end

   # app/models/reservation.rb
   class Reservation < ApplicationRecord
     belongs_to :user
     belongs_to :time_slot

     validates :user_id, uniqueness: { scope: :time_slot_id, message: "already reserved this slot" }
   end
   ```

2. **Time Slot Reservation Logic:**
   - Users can view available slots and reserve one.

   ```ruby
   # app/controllers/time_slots_controller.rb
   class TimeSlotsController < ApplicationController
     before_action :find_event

     def index
       @time_slots = @event.time_slots.where('start_time > ?', Time.now).order(:start_time)
     end

     def reserve
       @time_slot = @event.time_slots.find(params[:time_slot_id])

       if @time_slot.available_capacity > 0
         @time_slot.reservations.create!(user: current_user)
         redirect_to event_path(@event), notice: "You have successfully reserved the slot."
       else
         redirect_to time_slots_path(@event), alert: "No available capacity for this time slot."
       end
     end

     private

     def find_event
       @event = Event.find(params[:event_id])
     end
   end
   ```

3. **Views for Time Slot Reservation:**
   - Provide a simple UI where users can choose and reserve a time slot.

   ```erb
   <!-- app/views/time_slots/index.html.erb -->
   <h1>Available Time Slots</h1>
   <ul>
     <% @time_slots.each do |slot| %>
       <li>
         <%= slot.start_time.strftime('%I:%M %p') %> - <%= slot.end_time.strftime('%I:%M %p') %>
         (<%= slot.available_capacity %> seats left)
         <%= link_to 'Reserve', reserve_event_time_slot_path(@event, slot), method: :post if slot.available_capacity > 0 %>
       </li>
     <% end %>
   </ul>
   ```

4. **Admin Management for Time Slots:**
   - Admins can manage time slots via a simple CRUD interface.

   ```ruby
   # app/controllers/admin/time_slots_controller.rb
   class Admin::TimeSlotsController < ApplicationController
     def new
       @event = Event.find(params[:event_id])
       @time_slot = @event.time_slots.new
     end

     def create
       @event = Event.find(params[:event_id])
       @time_slot = @event.time_slots.new(time_slot_params)
       if @time_slot.save
         redirect_to admin_event_path(@event), notice: 'Time slot created successfully'
       else
         render :new
       end
     end

     private

     def time_slot_params
       params.require(:time_slot).permit(:start_time, :end_time, :capacity)
     end
   end
   ```

### 3. **Scaling Infrastructure:**
Ensure the system scales efficiently during high traffic events. Given your setup with AWS ECS Fargate and CloudFront, consider:
- **Auto-scaling:** Configure ECS to automatically scale tasks based on CPU/Memory usage or request counts.
- **CloudFront for Caching:** Cache static assets and reduce load on your app server.
- **Load Balancing:** Use ALB to distribute traffic evenly and avoid overload.

With these steps, you should have a robust system to handle high load during events by queuing users in a virtual waiting room and spreading traffic over reserved time slots.

To give users in the virtual waiting room a **time window** during which they can access the event, you can implement this feature by issuing time-limited access tokens. These tokens will expire after the allocated window, encouraging users to proceed without causing delays for others waiting in the queue.

Here’s how to implement a **time-limited access window** for virtual waiting room users in Rails:

### 1. **Modify the Virtual Waiting Room Service:**

Update your `VirtualWaitingRoomService` to include logic for assigning a specific time window to users who are allowed in.

#### Steps:

1. **Generate Access Tokens with Expiration:**
   - When a user is allowed into the event, generate an access token with a time-limited validity period (e.g., 5 minutes). This token will give users exclusive access to the event for that duration.

   ```ruby
   # app/services/virtual_waiting_room_service.rb
   class VirtualWaitingRoomService
     ACCESS_WINDOW_DURATION = 5.minutes

     def initialize(event_id)
       @event_id = event_id
       @redis = Redis.new
       @queue_key = "waiting_room:#{event_id}"
       @max_active_users = 100 # Set the max number of active users
     end

     def add_user_to_queue(user_id)
       @redis.lpush(@queue_key, user_id)
     end

     def allow_users_in
       active_users = @redis.lrange(@queue_key, 0, @max_active_users - 1)
       @redis.ltrim(@queue_key, @max_active_users, -1) # Keep remaining users in the queue
       active_users.each do |user_id|
         issue_access_token(user_id)
       end
       active_users
     end

     private

     def issue_access_token(user_id)
       access_token = SecureRandom.hex(16)
       expiration_time = Time.now + ACCESS_WINDOW_DURATION

       # Store the token with an expiration in Redis
       @redis.setex("user_access_token:#{user_id}", ACCESS_WINDOW_DURATION, access_token)

       # Optionally, send the token and expiration time to the user via WebSocket or email
       UserNotifier.send_access_details(user_id, access_token, expiration_time)
     end
   end
   ```

### 2. **Check Access Token Validity:**

When users try to access the event, check if their access token is valid and within the assigned time window. This can be done in the controller before granting them access.

```ruby
# app/controllers/events_controller.rb
class EventsController < ApplicationController
  before_action :validate_access_token, only: :show

  def show
    @event = Event.find(params[:id])
    # Proceed with event access
  end

  private

  def validate_access_token
    user_id = current_user.id
    provided_token = params[:access_token]
    stored_token = Redis.new.get("user_access_token:#{user_id}")

    if stored_token.nil? || stored_token != provided_token
      redirect_to waiting_room_path, alert: 'Your access window has expired or the token is invalid. Please wait again for your turn.'
    end
  end
end
```

### 3. **Front-End: Show Access Details and Countdown Timer:**

You can display the remaining time window to the user once they are admitted to the event, so they know how much time they have left to complete their action.

1. **Show access details to users:**
   When users are allowed into the event, show the access window and countdown timer.

   ```erb
   <!-- app/views/events/show.html.erb -->
   <div id="access-window-info">
     <p>You have been granted access to the event. Please complete your actions within the next <span id="remaining-time">5:00</span> minutes.</p>
   </div>

   <script>
     // Countdown timer
     const countdownElement = document.getElementById('remaining-time');
     let remainingTime = 300; // 5 minutes in seconds

     function updateTimer() {
       const minutes = Math.floor(remainingTime / 60);
       const seconds = remainingTime % 60;
       countdownElement.textContent = `${minutes}:${seconds.toString().padStart(2, '0')}`;
       remainingTime--;

       if (remainingTime < 0) {
         alert('Your access window has expired.');
         window.location.href = '<%= waiting_room_path %>'; // Redirect to waiting room
       }
     }

     setInterval(updateTimer, 1000);
   </script>
   ```

### 4. **Extend or Reissue Access (Optional):**

If you want to allow users to request an extension or reissue their access token under certain conditions (e.g., technical issues or slow loading), you can add logic to handle these cases.

```ruby
# app/controllers/waiting_room_controller.rb
class WaitingRoomController < ApplicationController
  def request_extension
    # Check if user is eligible for an extension (e.g., within a certain grace period)
    if eligible_for_extension?(current_user)
      VirtualWaitingRoomService.new(event_id).issue_access_token(current_user.id)
      redirect_to event_path(event_id), notice: "You have been granted additional time."
    else
      redirect_to waiting_room_path, alert: "You are not eligible for an extension at this time."
    end
  end

  private

  def eligible_for_extension?(user)
    # Define conditions for extending access (e.g., user has not recently extended)
    true
  end
end
```

### 5. **Scaling Considerations:**
Given that you’re preparing for high load, ensure the following:

1. **Efficient Redis Configuration:** Redis must be properly configured to handle the high volume of user data.
2. **Rate-limiting:** Implement rate-limiting on requests to the waiting room and token generation to prevent abuse.
3. **Autoscaling for Redis (if using AWS ElastiCache):** Ensure Redis can scale dynamically with traffic.
4. **Background Jobs for Access Management:** Use `Sidekiq` or `Delayed Job` to handle processing for admitting users and expiring tokens.

### 6. **Track Time Expiry via Background Jobs (Optional):**

Optionally, you can add a background job to track token expiration and notify users when their time is about to expire.

```ruby
class AccessWindowExpirationJob
  include Sidekiq::Worker

  def perform(user_id, event_id)
    # Notify user that their access window is expiring soon
    UserNotifier.notify_access_expiring(user_id)

    # Optionally, handle automatic revocation of access if needed
  end
end
```

With these changes, you can enforce a **time window** for users who are admitted to the event through the virtual waiting room, improving traffic control and user experience during high-demand events.

To handle both virtual waiting room users and reserved time slot users without exceeding the combined limit of 1000 active users, you need to coordinate the flow of users from both systems in a centralized way. Here's how to implement it:

### 1. **Track the Total Active Users:**

You'll need a central mechanism to track the total number of active users, combining both virtual waiting room users and reserved time slot users. Redis is ideal for fast, in-memory operations, and we can use it to keep count of the total active users.

#### Steps:

1. **Centralize User Count:**
   - Store the number of active users in Redis and ensure that both virtual waiting room and time slot users increment/decrement this count.
   
   ```ruby
   # app/services/active_user_tracker.rb
   class ActiveUserTracker
     MAX_ACTIVE_USERS = 1000

     def initialize
       @redis = Redis.new
       @active_users_key = "active_users_count"
     end

     def increment_active_users
       current_count = @redis.incr(@active_users_key)
       return current_count <= MAX_ACTIVE_USERS
     end

     def decrement_active_users
       @redis.decr(@active_users_key)
     end

     def active_users_count
       @redis.get(@active_users_key).to_i
     end

     def available_slots
       MAX_ACTIVE_USERS - active_users_count
     end
   end
   ```

2. **Integrate User Flow for Both Systems:**
   Modify both the virtual waiting room and time slot reservation systems to check and update the central active user count before allowing access.

### 2. **Modify Virtual Waiting Room Logic:**

Ensure that before admitting users from the virtual waiting room, the system checks if adding them would exceed the active user limit.

```ruby
# app/services/virtual_waiting_room_service.rb
class VirtualWaitingRoomService
  def allow_users_in
    active_user_tracker = ActiveUserTracker.new

    if active_user_tracker.available_slots > 0
      active_users = @redis.lrange(@queue_key, 0, active_user_tracker.available_slots - 1)
      @redis.ltrim(@queue_key, active_user_tracker.available_slots, -1) # Keep remaining users in the queue
      
      active_users.each do |user_id|
        if active_user_tracker.increment_active_users
          issue_access_token(user_id)
        else
          break
        end
      end

      active_users
    else
      [] # No slots available for users in the waiting room
    end
  end
end
```

### 3. **Modify Time Slot Reservation Logic:**

Ensure that before confirming a time slot reservation, the system checks the total active user count to avoid exceeding the limit.

```ruby
# app/controllers/time_slots_controller.rb
class TimeSlotsController < ApplicationController
  def reserve
    active_user_tracker = ActiveUserTracker.new
    @time_slot = @event.time_slots.find(params[:time_slot_id])

    if @time_slot.available_capacity > 0 && active_user_tracker.available_slots > 0
      if active_user_tracker.increment_active_users
        @time_slot.reservations.create!(user: current_user)
        redirect_to event_path(@event), notice: "You have successfully reserved the slot."
      else
        redirect_to time_slots_path(@event), alert: "No available capacity for new users."
      end
    else
      redirect_to time_slots_path(@event), alert: "No available capacity for this time slot or the system is full."
    end
  end
end
```

### 4. **Decrement User Count on Exit or Timeout:**

When a user’s session ends, either by exiting the event or timing out, decrement the active user count in both systems.

1. **Virtual Waiting Room Timeout:**
   - Use a background job to check when users' access tokens expire and remove them from the active users count.

   ```ruby
   class AccessWindowExpirationJob
     include Sidekiq::Worker

     def perform(user_id)
       # Notify user about timeout or remove access
       ActiveUserTracker.new.decrement_active_users
     end
   end
   ```

2. **Time Slot Expiration:**
   - Once a user has completed their reservation or the time slot has passed, remove them from the active user count.

   ```ruby
   # app/controllers/time_slots_controller.rb
   class TimeSlotsController < ApplicationController
     def complete_reservation
       # Mark user as done and free up their active user slot
       ActiveUserTracker.new.decrement_active_users
     end
   end
   ```

### 5. **Monitor and Limit Access Dynamically:**

Use Redis or other real-time monitoring tools to dynamically adjust the number of users allowed into the system. You can set up alerts if the active user count approaches the limit and block new admissions until there’s availability.

### 6. **Testing and Monitoring:**

Ensure you have proper logging and monitoring to track active users and the flow of admissions. You can set up CloudWatch alarms if you are using AWS to notify you if the active user count approaches the maximum.

```ruby
class MonitoringService
  def initialize
    @active_user_tracker = ActiveUserTracker.new
  end

  def log_active_users
    Rails.logger.info "Current Active Users: #{@active_user_tracker.active_users_count}"
  end
end
```

### Conclusion:

By centralizing the active user count and controlling admission from both the virtual waiting room and time slot reservation system, you can ensure that the combined total of active users never exceeds the set limit of 1000.

### PR DESCRIPTION
**Pull request by theachoem**
## Changes

image

## Related PR + demo:
- https://github.com/bookmebus/cm-market-app/pull/1913

### CODE REVIEW

File: app/interactors/spree_cm_commissioner/server_instances_bulk_creator.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      )
+
+      tasks_response = ecs_client.list_tasks(
+        cluster: ENV.fetch('AWS_CLUSTER_NAME'),
+        launch_type: 'FARGATE'
+      )
+
+      context.servers = []
+      context.task_arns = tasks_response.task_arns
+      context.task_arns.each do |task_arn|
```
Comment (channainfo) [inline app/interactors/spree_cm_commissioner/server_instances_bulk_creator.rb:20]: In the auto-scaling env, the number of servers can go up and down frequently. Saving them may cause lot of DB write operation. Can we just return the number of running tasks. If we need the read it back we can save it to our REDIS server or memcache to cache it for 2mins.

File: app/interactors/spree_cm_commissioner/waiting_guests_caller.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    def mark_as(full:)
+      firestore.col('waiting_rooms').doc('lobby').set({ full: full })
+    end
+
+    def firestore
+      context.firestore ||= Google::Cloud::Firestore.new(project_id: service_account[:project_id], credentials: service_account)
+    end
+
+    def service_account
+      context.service_account ||= Rails.application.credentials.cloud_firestore_service_account
```
Comment (channainfo) [inline app/interactors/spree_cm_commissioner/waiting_guests_caller.rb:52]:
```
@ervice_account ||= Rails.application.credentials.cloud_firestore_service_account
```

File: db/migrate/20241119032456_create_cm_waiting_room_sessions.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      t.string :jwt_token, null: false
+      t.string :remote_ip
+      t.datetime :expired_at, null: false
+
+      t.timestamps
+    end
+
+    add_index :cm_waiting_room_sessions, :guest_identifier, unique: true, if_not_exists: true
+    add_index :cm_waiting_room_sessions, :expired_at, if_not_exists: true
+    add_index :cm_waiting_room_sessions, :jwt_token, if_not_exists: true
```
Comment (channainfo) [inline db/migrate/20241119032456_create_cm_waiting_room_sessions.rb:14]: if we never query the record directly with the **jwt_token** field, better not index this field coz it contains a big string causing big storage. if we have to query directly with the field, we need create another field as a hash (md5/sha256) of the jwt_token and then index the hash instead.

File: app/controllers/spree/api/v2/storefront/accommodations_controller.rb [hunk 4]
```diff
@@ -3,6 +3,11 @@ module Api
     module V2
       module Storefront
         class AccommodationsController < ::Spree::Api::V2::ResourceController
+          # overrided
+          def required_waiting_room_session?
+            true
```
[the remark this answers is not in the document]
Comment (theachoem) [reply app/controllers/spree/api/v2/storefront/accommodations_controller.rb:8]: Currently only open it for specific API bong, which mean does not require authorization by default. API that required can use this.

File: app/interactors/spree_cm_commissioner/server_instances_bulk_creator.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  class ServerInstancesBulkCreator < BaseInteractor
+    def call
+      ecs_client = Aws::ECS::Client.new(
+        region: ENV.fetch('AWS_REGION'),
+        access_key_id: ENV.fetch('AWS_ACCESS_KEY_ID'),
+        secret_access_key: ENV.fetch('AWS_SECRET_ACCESS_KEY')
+      )
+
+      tasks_response = ecs_client.list_tasks(
+        cluster: ENV.fetch('AWS_CLUSTER_NAME'),
```
[the remark this answers is not in the document]
Comment (theachoem) [reply app/interactors/spree_cm_commissioner/server_instances_bulk_creator.rb:14]: But we want to record those tasks. I think it's seperate from the waiting itself.

**Pull request merged**